op-node/rollup/derive/blob_data_source.go

```diff
@@ -0,0 +1,174 @@
+		return ds.Next(ctx)
+	}
+	return data, nil
+}
+
+// open fetches and returns the blob or calldata (as appropriate) from all valid batcher
+// transactions in the referenced block. Returns an empty (non-nil) array if no batcher
+// transactions are found. Returns nil array whenever an error is returned.
+func (ds *BlobDataSource) open(ctx context.Context) ([]blobOrCalldata, error) {
+	_, txs, err := ds.fetcher.InfoAndTxsByHash(ctx, ds.ref.Hash)
+	if err != nil {
+		if errors.Is(err, ethereum.NotFound) {
+			return nil, NewResetError(fmt.Errorf("%w: failed to open blob data source", err))
+		}
+		return nil, NewTemporaryError(fmt.Errorf("%w: failed to open blob data source", err))
+	}
+
+	data, hashes := dataAndHashesFromTxs(txs, &ds.dsCfg, ds.batcherAddr)
+
+	if len(hashes) == 0 {
+		// there are no blobs to fetch so we can return immediately
+		return data, nil
+	}
+
+	// download the actual blob bodies corresponding to the indexed blob hashes
+	blobs, err := ds.blobsFetcher.GetBlobs(ctx, ds.ref, hashes)
+	if errors.Is(err, ethereum.NotFound) {
+		// If the L1 block was available, then the blobs should be available too. The only
+		// exception is if the blob retention window has expired, which we will ultimately handle
+		// by failing over to a blob archival service.
+		return nil, NewResetError(fmt.Errorf("%w: failed to fetch blobs", err))
+	} else if err != nil {
+		return nil, NewTemporaryError(fmt.Errorf("%w: failed to fetch blobs", err))
+	}
+
+	// go back over the data array and populate the blob pointers
+	if err := fillBlobPointers(data, blobs); err != nil {
+		// this shouldn't happen unless there is a bug in the blobs fetcher
+		return nil, NewResetError(fmt.Errorf("%w: failed to fill blob pointers", err))
+	}
+	return data, nil
+}
+
+// dataAndHashesFromTxs extracts calldata and datahashes from the input transactions and returns them. It
+// creates a placeholder blobOrCalldata element for each returned blob hash that must be populated
+// by fillBlobPointers after blob bodies are retrieved.
+func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address) ([]blobOrCalldata, []eth.IndexedBlobHash) {
+	data := []blobOrCalldata{}
+	var hashes []eth.IndexedBlobHash
+	blobIndex := 0 // index of each blob in the block's blob sidecar
+	for _, tx := range txs {
+		// skip any non-batcher transactions
+		if to := tx.To(); to == nil ||
+			*to != config.batchInboxAddress ||
+			!isValidBatchTx(tx, config.l1Signer, batcherAddr) {
+			blobIndex += len(tx.BlobHashes())
+			continue
+		}
+		// handle non-blob batcher transactions by extracting their calldata
+		if tx.Type() != types.BlobTxType {
+			calldata := eth.Data(tx.Data())
+			data = append(data, blobOrCalldata{nil, &calldata})
+			continue
+		}
+		// handle blob batcher transactions by extracting their blob hashes, ignoring any calldata.
+		if len(tx.Data()) > 0 {
+			log.Warn("blob tx has calldata, which will be ignored")
```

Comment:
Perhaps also include tx hash in the logline?

Reply:
done